stdlib/misc/sql_source.cpp

```diff
@@ -0,0 +1,234 @@
+      return;
+    }
+
+    // Setup connection to database
+    auto sql_config = args_.config();
+    LOG_IF(FATAL, sql_config.adapter() != "postgres")
+        << "Requested adapter " << sql_config.adapter()
+        << " does not exist. Only \"postgres\" is supported right now.";
+
+    try {
+      conn_.reset(new pqxx::connection{tfm::format(
+          "hostaddr=%s port=%d dbname=%s user=%s password=%s",
+          sql_config.hostaddr(), sql_config.port(), sql_config.dbname(),
+          sql_config.user(), sql_config.password())});
+    } catch (pqxx::pqxx_exception& e) {
+      LOG(FATAL) << e.base().what();
+    }
+  }
+
+  i64 total_elements() override {
+    try {
+    pqxx::work txn{*conn_};
+    // Count the number the number of groups
+    auto query = args_.query();
+    pqxx::row r = txn.exec1(tfm::format(
+        "SELECT COUNT(DISTINCT(%s)) FROM %s %s WHERE %s",
+        query.group(), query.table(), query.joins(), args_.filter()));
+    return r[0].as<i32>();
+    } catch (pqxx::pqxx_exception& e) {
+      LOG(FATAL) << e.base().what();
+    }
+  }
+
+  ElementArgs element_args_at(i64 element_idx) override {
+    proto::SQLElementArgs args;
+    args.set_filter(args_.filter());
+    size_t size = args.ByteSizeLong();
+
+    ElementArgs element_args;
+    element_args.args.resize(size);
+    args.SerializeToArray(element_args.args.data(), size);
+    element_args.row_id = element_idx;
+
+    return element_args;
+  }
+
+ private:
+  Result valid_;
+  scanner::proto::SQLEnumeratorArgs args_;
+  std::unique_ptr<pqxx::connection> conn_;
+};
+
+class SQLSource : public Source {
+ public:
+  SQLSource(const SourceConfig& config) :
+      Source(config), conn_(nullptr) {
+    bool parsed = args_.ParseFromArray(config.args.data(), config.args.size());
+    if (!parsed) {
+      RESULT_ERROR(&valid_, "Could not parse SQLSourceArgs");
+      return;
+    }
+
+    // Setup connection to database
+    auto sql_config = args_.config();
+    conn_.reset(new pqxx::connection{
+      tfm::format("hostaddr=%s port=%d dbname=%s user=%s password=%s",
+                  sql_config.hostaddr(), sql_config.port(), sql_config.dbname(), sql_config.user(), sql_config.password())
+    });
+
+    pqxx::work txn{*conn_};
+    pqxx::result types = txn.exec("SELECT oid, typname FROM pg_type");
+    for (auto row : types) {
+      pq_types_[row["oid"].as<pqxx::oid>()] = row["typname"].as<std::string>();
+    }
+  }
+
+  json row_to_json(pqxx::row row) {
+    json jrow;
+
+    for (pqxx::field field : row) {
+      std::string name(field.name());
+      if (name == "_scanner_id" || name == "_scanner_number") {
+        continue;
+      }
+
+      pqxx::oid type = field.type();
+      std::string& typname = pq_types_[type];
+      bool assigned = false;
+
+      #define ASSIGN_IF(NAME, TYPE) \
+        if (typname == NAME) { jrow[name] = field.as<TYPE>(); assigned = true; }
+
+      ASSIGN_IF("int4", i32);
+      ASSIGN_IF("float8", f32);
```

Comment:
Does this only support ints and floats? What about strings?

Reply:
Added support for more data types